src/lapack/backends/cusolver/cusolver_batch.cpp

```diff
@@ -79,251 +100,644 @@ void getri_batch(sycl::queue &queue, std::int64_t n, sycl::buffer<std::complex<d
+            cuDataType **a_batched = create_ptr_list_from_stride(a_, stride_a, batch_size);
+            CUDA_ERROR_FUNC(cuMemAlloc, cuda_result, &a_dev, sizeof(T *) * batch_size);
+            CUDA_ERROR_FUNC(cuMemcpyHtoD, cuda_result, a_dev, a_batched, sizeof(T *) * batch_size);
+
+            auto **a_dev_ = reinterpret_cast<cuDataType **>(a_dev);
+
+            CUSOLVER_ERROR_FUNC_T_SYNC(func_name, func, err, handle, get_cublas_fill_mode(uplo),
+                                       (int)n, a_dev_, (int)lda, nullptr, (int)batch_size);
+
+            free(a_batched);
+            cuMemFree(a_dev);
+        });
+    });
 }
-void ungqr_batch(sycl::queue &queue, std::int64_t m, std::int64_t n, std::int64_t k,
-                 sycl::buffer<std::complex<float>> &a, std::int64_t lda, std::int64_t stride_a,
-                 sycl::buffer<std::complex<float>> &tau, std::int64_t stride_tau,
-                 std::int64_t batch_size, sycl::buffer<std::complex<float>> &scratchpad,
-                 std::int64_t scratchpad_size) {
-    throw unimplemented("lapack", "ungqr_batch");
+
+// Scratchpad memory not needed as parts of buffer a is used as workspace memory
+#define POTRF_STRIDED_BATCH_LAUNCHER(TYPE, CUSOLVER_ROUTINE)                                      \
+    void potrf_batch(sycl::queue &queue, oneapi::mkl::uplo uplo, std::int64_t n,                  \
+                     sycl::buffer<TYPE> &a, std::int64_t lda, std::int64_t stride_a,              \
+                     std::int64_t batch_size, sycl::buffer<TYPE> &scratchpad,                     \
+                     std::int64_t scratchpad_size) {                                              \
+        return potrf_batch(#CUSOLVER_ROUTINE, CUSOLVER_ROUTINE, queue, uplo, n, a, lda, stride_a, \
+                           batch_size, scratchpad, scratchpad_size);                              \
+    }
+
+POTRF_STRIDED_BATCH_LAUNCHER(float, cusolverDnSpotrfBatched)
+POTRF_STRIDED_BATCH_LAUNCHER(double, cusolverDnDpotrfBatched)
+POTRF_STRIDED_BATCH_LAUNCHER(std::complex<float>, cusolverDnCpotrfBatched)
+POTRF_STRIDED_BATCH_LAUNCHER(std::complex<double>, cusolverDnZpotrfBatched)
+
+#undef POTRF_STRIDED_BATCH_LAUNCHER
+
+template <typename Func, typename T>
+inline void potrs_batch(const char *func_name, Func func, sycl::queue &queue,
+                        oneapi::mkl::uplo uplo, std::int64_t n, std::int64_t nrhs,
+                        sycl::buffer<T> &a, std::int64_t lda, std::int64_t stride_a,
+                        sycl::buffer<T> &b, std::int64_t ldb, std::int64_t stride_b,
+                        std::int64_t batch_size, sycl::buffer<T> &scratchpad,
+                        std::int64_t scratchpad_size) {
+    using cuDataType = typename CudaEquivalentType<T>::Type;
+
+    overflow_check(n, nrhs, lda, ldb, stride_a, stride_b, batch_size, scratchpad_size);
+
+    // cuSolver function only supports nrhs = 1
+#ifndef POTRS_BATCHED_MULTIPLE_NRHS_SUPPORTED
```

Comment:
There is C++ way of doing this, not involving preprocessor's macro (`if constexpr (not cusolver_batched_potrs_supports_multiple_nrhs) { if (nrhs != 1) { throw } }`).
Here it is also not clear from the code if it is supported, or not, without explicit definition.
Comment on line 346 got detached. Better to remove it as the code is quite self-documenting here.

Reply:
I think the issue with this approach is where is `cusolver_batched_potrs_supports_multiple_nrhs` going to reside? If this is to be set in a single place it would have to be a global variable, which is undesirable, as each of these are free functions.

Personally I think having this check done this way makes sense as we are using the legacy API. It seems very unlikely that NVIDIA will come out with an update eliminating this constraint for an API they are probably going to deprecate at some point. So any support to allow `nrhs > 1` would require a change to the code and therefore the check can be simply deleted at that time. However, I still do not have any issues with making a way to by pass it. Just noting that it is unlikely to be needed.

Reply:
> I think the issue with this approach is where is cusolver_batched_potrs_supports_multiple_nrhs going to reside? If this is to be set in a single place it would have to be a global variable, which is undesirable, as each of these are free functions.

This is not a variable, but a constant, leaving no trace in the binary. Could be declared at the beginning of the source.

> It seems very unlikely that NVIDIA will come out with an update eliminating this constraint for an API they are probably going to deprecate at some point. 

This is a good point. So we can omit on this expectation.
Documentation explicitly states `The routine will be removed in the next major release.` Why we using it here if it won't work in a year?

Reply:
I have removed the macro.
The reason to use the legacy API is that the new API is not yet fleshed out. There would be a lot of missing functions, I would expect that when the legacy API is deprecated they will also bring in more supported operations for the new API.

Reply:
> The reason to use the legacy API is that the new API is not yet fleshed out.

I skimmed over the spec and all API, having deprecation notices, also have a proposed new API to use. What is missed?

Reply:
If you take a look at the number of supported lapack operations, the 64-bit API does not support nearly as many operations. For example the 64-bit dense eigenvalue solver section only really contains two lapack operations. You can then see that the legacy API has many more. 
Legacy API eigenvalue solvers: https://docs.nvidia.com/cuda/cusolver/index.html#cuSolverDN-eigensolver-reference
64-bit API eigenvalue solvers: https://docs.nvidia.com/cuda/cusolver/index.html#cuSolverDN-eigensolver-reference-64bit

Reply:
Isn't this update only concerned with potrs and getrs which are being deprecated and have new API proposed?